Improve I/O performance with using /dev/mem or /dev/gpiomem

jdevelop opened the pull request:
According to some metrics, /dev/mem and /dev/gpiomem are much faster to access for pin I/O.
Periph has significant performance penalty for reading/writing GPIO via sysfs, which is order of magnitudes slower than direct access.

With simple edge detection/busy wait it is quite possible to use HX711 on Raspberry Pi Zero.

See #1 for more details.

jdevelop committed:
#1 - Use RPIO library

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

 Fixed pin modes.

MichaelS11 commented:
I think this is great, I am glad it is working on Raspberry Pi Zero using go-rpio/v4.

I would like to keep this repo's master branch still using periph.io. My hope is sometimes they will updated GPIO so it will work better/faster like it should. Happy to add another branch, maybe called rpio.

jdevelop commented:
I guess that it will be more convenient if we can do the following:
- restructure the repo and move the low-level GPIO access functions to some abstractions. So there will be a few high-level functions like ReadData, ReadDataRaw, ReadMedian, Calibrate etc, and low level interface to actually interact with pins. Not sure about the performance penalty here, with the indirect functions call chain - but perhaps the compiler is smart enough to inline the nested calls.
- provide build tags to build for Periph and RPIO-related sources in that low-level access package.

in that case it would be easier for people to use whatever they want, and once Periph will improve the GPIO access - it would be possible to just drop the RPIO modules without breaking the high-level contract.

Makes sense?

MichaelS11 commented:
Tags sounds good. Have it default to periph.io and add a tag rpio for it to switch to using go-rpio

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Added Periph and rpio modes.

jdevelop commented:
@MichaelS11 updated, pls have a look. I kept the interfaces _slightly_ incompatible to make sure people will use the version they expect.

MichaelS11 commented on go.mod:
```diff
@@ -0,0 +1,8 @@
+require (
+	github.com/stianeikeland/go-rpio/v4 v4.4.0
+	periph.io/x/periph v3.6.2+incompatible
+)
```
Don't know if we really need the go.mod file

jdevelop replied:
Better to have that one for the people who use modules, doesn't hurt if used with `$GOPATH`

At least if periph goes up one version - there won't be any surprises.

MichaelS11 commented on hx711_rpio.go:
```diff
@@ -0,0 +1,343 @@
+	"github.com/stianeikeland/go-rpio/v4"
+)
+
+var timeoutError = fmt.Errorf("timeout")
```
This should be ErrTimeout

jdevelop replied:
Fixed

MichaelS11 reviewed (changes requested):
Would rather keep the api the same.

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Simplify code, keep IO-agnostic methods in the separate file.

jdevelop reviewed (commented):
Updated interface and separate modules.

MichaelS11 commented on go.mod:
```diff
@@ -0,0 +1,8 @@
+module github.com/MichaelS11/go-hx711
+
+go 1.13
```
Should this be go 1.12 or even lower?

jdevelop replied:
Could be 1.11 perhaps, but doesn't really matter.

MichaelS11 commented on hx711_rpio.go:
```diff
@@ -0,0 +1,128 @@
+// NewHx711 creates new Hx711.
+// Make sure to set clockPin and dataPin to the correct pins.
+// https://cdn.sparkfun.com/datasheets/Sensors/ForceFlex/hx711_english.pdf
+func NewHx711(clockPinStr string, dataPinStr string) (*Hx711, error) {
```
Maybe change this to clockPinNum and dataPinNum?
Also add command and link about how the pin number works.
https://godoc.org/github.com/stianeikeland/go-rpio#Pin

MichaelS11 commented on hx711_rpio.go:
```diff
@@ -0,0 +1,128 @@
+// Make sure to set clockPin and dataPin to the correct pins.
+// https://cdn.sparkfun.com/datasheets/Sensors/ForceFlex/hx711_english.pdf
+func NewHx711(clockPinStr string, dataPinStr string) (*Hx711, error) {
+	clockPin, err := strconv.ParseInt(clockPinStr, 10, 32)
```
Should this be 64 instead of 32?

jdevelop replied:
I don't think that 64 bit is a requirement for GPIO pin numbering.

MichaelS11 commented on hx711_rpio.go:
```diff
@@ -0,0 +1,128 @@
+	}
+	for i := 0; i < 200000; i++ {
+		if !hx711.dataPin.EdgeDetected() {
+			time.Sleep(5 * time.Microsecond)
```
Don't think there is a need to have this tight of a loop. We are working in Milliseconds not Microseconds.
The small sleep times are going to do a CPU busy wait, which creates a lot of CPU load.

jdevelop replied:
According to the [datasheet](https://cdn.sparkfun.com/datasheets/Sensors/ForceFlex/hx711_english.pdf) page 5 - `T2` value is in the range of the range 0.1 microseconds, so 5 is pretty much a giveaway.

MichaelS11 replied:
You are correct, my bad.

MichaelS11 replied:
From what I found (see note in the code) it takes around 80 milliseconds for the chip to get ready. Did you find something different?

MichaelS11 commented on hx711_rpio.go:
```diff
@@ -0,0 +1,128 @@
+		return nil
+	}
+	for i := 0; i < 200000; i++ {
+		if !hx711.dataPin.EdgeDetected() {
```
Can switch this around. Do if hx711.dataPin.EdgeDetected() then return. Then can remove the else.

MichaelS11 reviewed (changes requested):
Could you add something to the readme as well?

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Add documentation, apply code review suggestions.

jdevelop commented:
Updated README, lowered Go version requirements

MichaelS11 commented on README.md:
````diff
@@ -22,6 +22,23 @@ Side note, in my testing using 3V input had better consistency then using a 5V i
+
+```
+go build -tags=rpio 
+```
````
I think that is a lot of information for most people to go through. What do you think about splitting it in two? A very short thing about tags, then details at the very end of the readme?

jdevelop replied:
please take a look at the update.

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Updated documentation.

MichaelS11 commented on hx711_rpio.go:
```diff
@@ -0,0 +1,129 @@
+		if hx711.dataPin.EdgeDetected() {
+			return nil
+		}
+		time.Sleep(5 * time.Microsecond)
```
From what I found (see note in the code) it takes around 80 milliseconds for the chip to get ready. Did you find something different?

Still wondering if 5 microsecond sleep is to small.

jdevelop replied:
I was testing with different timeouts, my findings are:
- anything above 1 millisecond leads to unstable reading results ( about 5 - 10% errors )
- with 2 microseconds the number of iterations before the edge is detected is 27000-31000, which gives 27-30 milliseconds delay

I don't really know *why* setting up the delay of 1 millisecond causes errors. My suspicion is that 10 milliseconds is not really 1 millisecond.

The most stable and reasonable results that I've got were with 250 microseconds to wait until trying to read GPIO memory to see whether there was falling edge detected. The number of iterations is less than 500, but I feel like it could be a good idea to cap it to 1500, just in case.

Hope that helps :)

MichaelS11 replied:
Thank you so much, that is great :)

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Rename `rpio` flag to more descriptive `gpiomem` flag. Fixed error with
gpio compilation.

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Fixed timings, updated constants.

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Use microseconds instead of milliseconds.

MichaelS11 commented on hx711_rpio.go:
```diff
@@ -0,0 +1,132 @@
+// The pin numbers must comply with BCM numbering schema.
+// https://cdn.sparkfun.com/datasheets/Sensors/ForceFlex/hx711_english.pdf
+// https://godoc.org/github.com/stianeikeland/go-rpio#Pin
+func NewHx711(clockPinStr string, dataPinStr string) (*Hx711, error) {
```
Was there a reason why you changed this back to Str from Num?

jdevelop replied:
Compilation :) I forgot to change the names of variables and with the build tags I missed this.

Should be nice to setup travis build that could build everything with all tags, and may be even run some tests.

MichaelS11 replied:
Still showing Str

jdevelop replied:
Yeah, fixed it.

jdevelop committed:
#1 - Improve edge detection, use shorter wait intervals.

Rename pin parameter names.

MichaelS11 reviewed (approved):
Thank you!
